tracer/src/emulator/cpu.rs

```
@@ -1391,6 +1441,49 @@ impl Cpu {
     pub fn get_mut_terminal(&mut self) -> &mut Box<dyn Terminal> {
         self.mmu.get_mut_uart().get_mut_terminal()
     }
+
+    fn handle_jolt_cycle_marker(&mut self, ptr: u32, event: u32) -> Result<(), Trap> {
+        match event {
+            JOLT_CYCLE_MARKER_START => {
+                let label = self.read_c_string(ptr)?; // guest NUL-string
+                self.active_markers.insert(
+                    ptr,
+                    ActiveMarker {
+                        label,
+                        start_instrs: self.executed_instrs,
+                        start_trace_len: self.trace.len(),
+                    },
+                );
+            }
+
+            JOLT_CYCLE_MARKER_END => {
+                if let Some(mark) = self.active_markers.remove(&ptr) {
+                    let real = self.executed_instrs - mark.start_instrs;
+                    let virt = self.trace.len() - mark.start_trace_len;
+                    println!(
+                        "\"{}\": {} RV32IM cycles, {} virtual cycles",
+                        mark.label, real, virt
+                    );
+                }
+            }
+            _ => { /* ignore other opcodes */ }
```

let's panic here with "unexpected event"